packages/react-reconciler/src/ReactFiberWakeable.new.js

```diff
@@ -48,3 +61,24 @@ export function throwIfInfinitePingLoopDetected() {
     // the render phase so that it gets the component stack.
   }
 }
+
+export function trackSuspendedThenable<T>(
+  thenable: Thenable<T>,
+  index: number,
+) {
+  if (suspendedThenables === null) {
+    suspendedThenables = [];
+  }
+  suspendedThenables[index] = thenable;
```

Comment:
So this is a holey array which is pretty deopted. I wonder if it might be worth just setting it to null for every `use()` call even if nothing suspends. Perhaps reusing the same array every time and filling it with null each time. Maybe that's worse.

Reply:
Maybe @gsathya would know which is best

Reply:
Looks like this got replaced by expandos? Expandos will cause hidden class transition leading to a potential deopt. Curious if WeakMap works for you? Or is too slow? 

I guess, the tradeoff here is between slow lookup in WeakMaps or a deopt in other parts that use this thenable -- I don't know which is the more common workload.

As an aside, if looking up the Promise value synchronously is the common case, then we could champion a `Promise.prototype.inspect` proposal to add this to the Promise API.

Reply:
`Promise.prototype.inspect` wouldn't help in this particular case because this array is for tracking thenables that aren't cached. We use it when we know two thenables represent the same data even though they are different objects. For example;

```js
const helloPromise = Promise.resolve('Hello');

function HelloWorld() {
  // This component uses different promises on every render attempt,
  // but if no React state has changed we know that the underlying value
  // can be reused.
  return
    // This promise is the same every time, so we can read the value
    // from the expando (a la Promise.prototype.inspect)
    use(helloPromise) +
    // This promise is new every time, but we know it represents the
    // same thing, so we reuse the previous value from the array.
    use(Promise.resolve('World');
}
```

In the current implementation, this example would produce a sparse array with a value at index 1 and a hole at index 0.

The reason the first one is is because if the promise can be inspected, there's no reason to track it. So Seb's suggestion is that it might be better to push to the array even if it's not needed.

There are other solutions to avoid a holey array, like only advancing the thenableIndex if the promise hasn't resolved. I think that would require one additional null check to each read of an inspectable promise, but it might be worth it to prevent a deopt.